### PR TITLE
emacs: add an option to install the C source

### PR DESCRIPTION
**Pull request by dudebout**
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Install the source files, and the associated TAGS files. Use
.dir-locals.el to select the TAGS file automatically.

Turn the option on by default, to get a consistent behavior when using
xref-find-definitions (M-.), whether looking for an elisp (e.g. defun)
or a C (e.g. defalias) definition.

### CODE REVIEW

**Comment by joachifm** [comment]
How does this compare to https://github.com/NixOS/nixpkgs/pull/17273

**Comment by dudebout** [comment]
Same idea, different implementation. Advantages of the implementation in this PR:
* puts the C source at the same level as the associated elisp source
* does not store C source inside a site-lisp directory
* uses the upstream `tags` Makefile target instead of reimplementing it
* does not use the emacs extension mechanism (package.el / emacsWithPackages) to present a core element of emacs
* does not create an artificial version number (1.0.0)
* does not set the `source-directory` variable since its documented purpose is different: `Directory in which Emacs sources were found when Emacs was built.`
* uses `.dir-locals.el` files to automatically recognize the TAGS file in the source directories

**Comment by DamienCassou** [comment]
As @joachifm proposed in #17273, what about a separate output? If this PR is getting merged, should the solution in #17273 be reverted? /cc @binarin

**Comment by binarin** [comment]
Initial version of #17273 was almost  the same as this PR (sans separate patch file). And the main disadvantage of that approach is the need to rebuild the whole emacs just to get sources (as you can't use prebuilt package from binary cache).  But it doesn't matter for me, as I'm always rebuilding my emacs with different options. So if this PR is accepted, then #17273 should be reverted.

File: pkgs/applications/editors/emacs/default.nix [hunk 1]
```diff
@@ -31,7 +32,8 @@ stdenv.mkDerivation rec {
     sha256 = "0cwgyiyymnx4xdg99dm2drfxcyhy2jmyf0rkr9fwj9mwwf77kwhr";
   };
 
-  patches = lib.optional stdenv.isDarwin ./at-fdcwd.patch;
+  patches = lib.optional stdenv.isDarwin ./at-fdcwd.patch
+    ++ lib.optional withCsrc ./install-C-src.patch;
```
Comment (binarin) [inline pkgs/applications/editors/emacs/default.nix]: Inline patches are frowned upon. And in this case it's not worth it, just do the same thing somewhere in `postInstall`.

**Review by binarin — CHANGES_REQUESTED** [review]
And `Directory in which Emacs sources were found when Emacs was built` will not hold anyway, as build happens in temporary directory. One possible idea is to replace `fetchurl` with `fetchzip` and point `default-directory` directly to nix store.

File: pkgs/applications/editors/emacs/default.nix (hunk 1 above)
Comment (dudebout) [inline pkgs/applications/editors/emacs/default.nix]: @binarin, thanks for pointing this out. Is this documented somewhere or just general knowledge? Happy to change it.

The main advantage of doing it in the Makefile is to have access to the version number. I debated whether it was better to add the version number in the derivation or to do it this way.

Comment (dudebout) [inline pkgs/applications/editors/emacs/default.nix]: @binarin I just remembered one advantage of the patch. It is a lot more self contained. Without that you need to pass some target to make and do things in `postInstall`. It sprinkles the option around a little bit, but not that bad.

**Comment by dudebout** [comment]
@DamienCassou What is the advantage of putting the C source in a separate output? The elisp source is in `out`.

Suppose we want to put the C source in the output `csrc`. How do we nicely integrate things in `site-start.el`? I don't think the builder script can know if we are installing `emacs` or `emacs.csrc`. It needs to build both as the steps are intertwined. Therefore, we cannot conditionally set `find-function-C-source-directory`. So we would always have to set a reference from `emacs` to `emacs.csrc`that could be invalidated by GC.

@binarin I don't understand the need to rebuild you mention. With this PR, when you install emacs, you automatically install the matching C source. Same as with the gzipped elisp source. As a side note, I looked into gzipping the C source but `help-C-file-name` relies on the file name ending in `.c`.

@binarin The full documentation for `source-directory` does mention that the data might not still be there:
```
Directory in which Emacs sources were found when Emacs was built.
You cannot count on them to still be there!
```
There is no need to set this variable to something "better" for setting `find-function-C-source-directory`; it is only used as a default value.

**Comment by DamienCassou** [comment]
@dudebout no idea :-). I liked the external package solution because it means that both people who want the C sources and those who don't want them can benefit from a binary Emacs in the nix cache. But I don't care enough :-).

**Comment by dudebout** [comment]
I see. The binary cache issue is that only one option combination can be in the cache.

File: pkgs/applications/editors/emacs/default.nix (hunk 1 above)
Comment (binarin) [inline pkgs/applications/editors/emacs/default.nix]: I was told about inline patches in some other PR - that they should be downloadable from external source. And now I think that this is indeed good policy - it pushes you to improving upstream build process and making it more universal =)

Comment (dudebout) [inline pkgs/applications/editors/emacs/default.nix]: Reimplemented in the derivation.

**Comment by dudebout** [comment]
I have rewritten the functionality without the patch to the Makefile. I do not see an easy way to make a separate output at this time (I will keep looking). Anything else you guys need from me?

**Comment by DamienCassou** [comment]
Thanks for your work. What about removing the `emacs-source-directory` package?

**Comment by dudebout** [comment]
Ah yes! Done.

File: pkgs/build-support/emacs/wrapper.nix [hunk 2]
```diff
@@ -85,13 +85,14 @@ stdenv.mkDerivation {
      done
 
      siteStart="$out/share/emacs/site-lisp/site-start.el"
+     siteStartByteCompiled="$siteStart"c
```
Comment (binarin) [inline pkgs/build-support/emacs/wrapper.nix]: trailing `c`

**Review by binarin — CHANGES_REQUESTED** [review]

**Comment by edolstra** [comment]
What exactly is the use case for making the source code available as an output? AFAIK we don't do that for any other package.

**Comment by dudebout** [comment]
@binarin that trailing c is taking a .el and making a .elc. That is the
expected behavior.

**Comment by DamienCassou** [comment]
@edolstra I'm not sure I understand the question. I don't know if putting the sources in a separate output is any useful except maybe to maximize reuse of the Emacs binary in the nix cache for people wanting the C sources and people not wanting the C sources.

**Comment by edolstra** [comment]
My point is that you can already get sources by doing `nix-build '<nixpkgs>' -A emacs.src`, the same as any other package.

**Comment by DamienCassou** [comment]
@edolstra For emacs, it's good to have the sources packaged because Emacs can show its own sources when the user asks for them. If there is another way to give access to sources to Emacs, can you please help us?

**Comment by dudebout** [comment]
`emacs.src` will give you the tarball containing all the sources plus licence, Makefile etc. What we are after here is a small subset of it, just the C source, untarred, with some configuration to let emacs expose it to the user seamlessly. With this in place, you just install emacs, and with one keybinding (M-., if you were curious ;-)) you can find the location of any lisp or C definition. The lisp part was already there, and we are just putting the C part in place.

**Comment by dudebout** [comment]
Should we go ahead with this one, and we can revisit creating a "source" output later?

**Comment by DamienCassou** [comment]
LGTM. +1 for merging.

**Comment by dudebout** [comment]
Any objection?

**Comment by DamienCassou** [comment]
I think your code has waited for too long. I'm merging it. Thank you very much.

**Pull request merged**